Makefile

```diff
@@ -1,102 +1,102 @@
-BAZEL_CONFIG_TSAN = # no working config
+LOCAL_OS := $(shell uname)
+ifeq ($(LOCAL_OS),Linux)
+    TARGET_OS ?= linux
+    READLINK_FLAGS="-f"
+else ifeq ($(LOCAL_OS),Darwin)
+    TARGET_OS ?= darwin
+    READLINK_FLAGS=""
+else
+    $(error This system's OS $(LOCAL_OS) isn't supported)
 endif
 
-build:
-	export PATH=$(PATH) CC=$(CC) CXX=$(CXX) && bazel $(BAZEL_STARTUP_ARGS) build $(BAZEL_BUILD_ARGS) $(BAZEL_CONFIG_DEV) $(BAZEL_TARGETS)
+export TARGET_OUT ?= $(shell pwd)/out/$(TARGET_ARCH)_$(TARGET_OS)
 
-build_envoy:
-	export PATH=$(PATH) CC=$(CC) CXX=$(CXX) && bazel $(BAZEL_STARTUP_ARGS) build $(BAZEL_BUILD_ARGS) $(BAZEL_CONFIG_REL) //src/envoy:envoy
+ifeq ($(BUILD_WITH_CONTAINER),1)
+export TARGET_OUT = /work/out/$(TARGET_ARCH)_$(TARGET_OS)
+CONTAINER_CLI ?= docker
+DOCKER_SOCKET_MOUNT ?= -v /var/run/docker.sock:/var/run/docker.sock
+IMG ?= gcr.io/istio-testing/build-tools:2019-10-11T13-37-52
+UID = $(shell id -u)
+PWD = $(shell pwd)
 
-clean:
-	@bazel clean
+$(info Building with the build container: $(IMG).)
 
-test:
-	export PATH=$(PATH) CC=$(CC) CXX=$(CXX) && bazel $(BAZEL_STARTUP_ARGS) build $(BAZEL_BUILD_ARGS) $(BAZEL_CONFIG_DEV) //src/envoy:envoy
-	export PATH=$(PATH) CC=$(CC) CXX=$(CXX) && bazel $(BAZEL_STARTUP_ARGS) test $(BAZEL_BUILD_ARGS) $(BAZEL_CONFIG_DEV) $(BAZEL_TARGETS)
-	GO111MODULE=on go test ./...
+# Determine the timezone across various platforms to pass into the
+# docker run operation. This operation assumes zoneinfo is within
+# the path of the file.
+TIMEZONE=`readlink $(READLINK_FLAGS) /etc/localtime | sed -e 's/^.*zoneinfo\///'`
 
-test_asan:
-	export PATH=$(PATH) CC=$(CC) CXX=$(CXX) && bazel $(BAZEL_STARTUP_ARGS) build $(BAZEL_BUILD_ARGS) $(BAZEL_CONFIG_ASAN) //src/envoy:envoy
-	export PATH=$(PATH) CC=$(CC) CXX=$(CXX) && bazel $(BAZEL_STARTUP_ARGS) test $(BAZEL_BUILD_ARGS) $(BAZEL_CONFIG_ASAN) -- $(BAZEL_TARGETS) $(SANITIZER_EXCLUSIONS)
-	GO111MODULE=on go test ./...
+RUN = $(CONTAINER_CLI) run -t -i --sig-proxy=true -u $(UID):docker --rm \
+	-e IN_BUILD_CONTAINER="$(BUILD_WITH_CONTAINER)" \
+	-e TZ="$(TIMEZONE)" \
+	-e TARGET_ARCH="$(TARGET_ARCH)" \
+	-e TARGET_OS="$(TARGET_OS)" \
+	-e TARGET_OUT="$(TARGET_OUT)" \
+	-e HUB="$(HUB)" \
+	-e TAG="$(TAG)" \
+	-v /etc/passwd:/etc/passwd:ro \
+	$(DOCKER_SOCKET_MOUNT) \
+	$(CONTAINER_OPTIONS) \
+	--mount type=bind,source="$(PWD)",destination="/work" \
+	--mount type=volume,source=go,destination="/go" \
+	--mount type=volume,source=gocache,destination="/gocache" \
+	-w /work $(IMG)
 
-test_tsan:
-	export PATH=$(PATH) CC=$(CC) CXX=$(CXX) && bazel $(BAZEL_STARTUP_ARGS) build $(BAZEL_BUILD_ARGS) $(BAZEL_CONFIG_TSAN) //src/envoy:envoy
-	export PATH=$(PATH) CC=$(CC) CXX=$(CXX) && bazel $(BAZEL_STARTUP_ARGS) test $(BAZEL_BUILD_ARGS) $(BAZEL_CONFIG_TSAN) --test_env=TSAN_OPTIONS=suppressions=$(TOP)/tsan.suppressions -- $(BAZEL_TARGETS) $(SANITIZER_EXCLUSIONS)
-	GO111MODULE=on go test ./...
+MAKE = $(RUN) make --no-print-directory -e -f Makefile.core.mk
```

Comment:
Note: this still breaks `PATH` in the `BUILD_WITH_CONTAINER=1 make build` builds that fail immediately.

Reply:
Why do we care about all host environment variables to be passed to docker? Perhaps we could explicitly pass those that are needed?

Alternatively, could you replace the `PATH` in `Makefile.core.mk` with something LLVM-specific, and prepend it in make targets, like I suggested earlier?

Reply:
I'm not trying to get BUILD_WITH_CONTAINER=1 to work at all with this PR. I'm just trying to get the common files on par with all the other repos. After that, I/we can start working on container-based builds.

Reply:
Sounds good.